[X86] Elect to tail call when `sret` ptr is passed to the callee

antoniofrighetto opened the pull request:
We may be able to allow the callee to be tail-called when the caller expects a `sret` pointer argument, as long as this pointer is forwarded to the callee.

Fixes: https://github.com/llvm/llvm-project/issues/146303.

rnk reviewed (changes requested):
Thanks, the test cases suggest that this would be a major improvement.

rnk reviewed (approved):
Looks good! Sorry for the delay, I went on summer vacation.

antoniofrighetto committed:
[X86] Precommit tests for PR146575 (NFC)

antoniofrighetto committed:
[X86] Elect to tail call when `sret` ptr is passed to the callee

We may be able to allow the callee to be tail-called when the caller
expects a `sret` pointer argument, as long as this pointer is forwarded
to the callee.

Fixes: https://github.com/llvm/llvm-project/issues/146303.
